server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -330,28 +344,33 @@ public void run() {
                 var builder = new UpdateHealthInfoCacheAction.Request.Builder().nodeId(clusterService.localNode().getId());
                 healthProgresses.forEach(changedHealthInfo -> changedHealthInfo.updateRequestBuilder(builder));
 
-                var healthNodeId = lastSeenHealthNode.get();
                 var listener = ActionListener.<AcknowledgedResponse>wrap(response -> {
-                    // Don't update the latest health info if the health node has changed while this request was being processed.
-                    if (Objects.equals(healthNodeId, lastSeenHealthNode.get()) == false) {
-                        return;
+                    // Only record health progress if this monitoring instance hasn't been cancelled in the meantime.
+                    // This avoids any unwanted writes to the HealthTrackers' states after a new monitoring instance has possibly
+                    // already started.
+                    if (cancelled == false) {
+                        healthProgresses.forEach(HealthTracker.HealthProgress::recordProgressIfRelevant);
                     }
-                    healthProgresses.forEach(HealthTracker.HealthProgress::recordProgressIfRelevant);
                 }, e -> {
                     if (e.getCause() instanceof NodeNotConnectedException || e.getCause() instanceof HealthNodeNotDiscoveredException) {
                         logger.debug("Failed to connect to the health node [{}], will try again.", e.getCause().getMessage());
                     } else {
                         logger.debug(() -> format("Failed to send health info to health node, will try again."), e);
                     }
                 });
-                client.execute(UpdateHealthInfoCacheAction.INSTANCE, builder.build(), ActionListener.runAfter(listener, scheduleNextRun));
+                client.execute(
+                    UpdateHealthInfoCacheAction.INSTANCE,
+                    builder.build(),
+                    ActionListener.runAfter(listener, releaseAndScheduleNextRun)
+                );
                 nextRunScheduled = true;
             } catch (Exception e) {
                 logger.warn(() -> format("Failed to run scheduled health monitoring on thread pool [%s]", executor), e);
+                // Make sure to release the "lock" in case something goes wrong, so we can try again in the next iteration.
             } finally {
                 // If the next run isn't scheduled because for example the health info hasn't changed, we schedule it here.
                 if (nextRunScheduled == false) {
-                    scheduleNextRun.run();
+                    releaseAndScheduleNextRun.run();
```

Comment:
I don't think this is correct here. 

This finally block might run **before** our action listener https://github.com/elastic/elasticsearch/pull/105674/files#diff-ec8692646c2c3b00102070f6daff810ef1e176024d23206287684849dd17ace2R364

IIUC it will then set `inFlight` to `false`. 

I don't think juggling `runAfter` and `RunOnce` helps readability either but perhaps that's a personal preference.

Reply:
When `finally` block runs, given there is no exception, `nextRunScheduled` would have been set to `true` so it will skip running the `releaseAndScheduleNextRun`. 

Does this address your concern?

Reply:
Ah yes! Thanks Mary, that makes sense.